File: 781. Rabbits in Forest/numRabbits.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    int numRabbits(vector<int>& answers) {
        int ans = 0;
        int max = 0;
        for(auto num : answers){
            if(num > max){
                max = num;
            }
        }
        vector<int> mp(max+1);

        for(auto num : answers){
            mp[num] += 1;
        }

        for(int i=0 ;i<mp.size(); i++){
            if(mp[i] % (i+1) == 0){
                ans += mp[i];
            }
            else{
                int cnt = mp[i] / (i+1);
                ans += ((i+1) * (cnt+1));
            }
        }
        return ans;
    }
};

File: 781. Rabbits in Forest/numRabbits.js
/**
 * @param {number[]} answers
 * @return {number}
 */
var numRabbits = function (answers) {
    let ans = 0
    let mp = new Map()

    for (let rab of answers) {
        if (!mp.has(rab)) {
            mp.set(rab, 0)
        }
        mp.set(rab, mp.get(rab) + 1)
    }

    //  SECOND SOLUTION #2
    for (let [key] of mp) {
        if (mp.get(key) % (key + 1) === 0) {
            ans += mp.get(key)
        }
        else {
            let cnt = Math.floor(mp.get(key) / (key + 1)) // => there is cnt+1 PAIRS of same color rabbits
            ans += ((key + 1) * (cnt + 1))
        }
    }
    return ans;
};


//  FIRST SOLUTION => come to mind #1
// for(let i=0 ;i<answers.length; i++){
//     let get = mp.get(answers[i])
//     let num = 1 + answers[i]
//     if(get >= num){
//         ans+= num
//         mp.set(answers[i] , get-num)
//     }
// }

// for(let [key] of mp){
//     if(mp.get(key) > 0){
//         ans += key+1
//     }
// }